crates/block-producer/src/metrics.rs

```diff
@@ -0,0 +1,205 @@
+use std::collections::HashMap;
+
+use gw_common::{CKB_SUDT_SCRIPT_ARGS, H256};
+use gw_otel::metric::{
+    counter::Counter,
+    encoding::text::Encode,
+    family::Family,
+    gauge::Gauge,
+    prometheus_client,
+    registry::{Registry, Unit},
+    Lazy, OnceCell,
+};
+use gw_store::{traits::chain_store::ChainStore, Store};
+use gw_types::prelude::Unpack;
+use serde::Deserialize;
+use tracing::instrument;
+
+pub const ENV_METRIC_MONITOR_CUSTODIAN_ENABLE: &str = "METRIC_MONITOR_CUSTODIAN_ENABLE";
+pub const ENV_METRIC_MONITOR_CUSTODIAN_VEC_JSON: &str = "METRIC_MONITOR_CUSTODIAN_VEC_JSON";
+
+pub static BP_METRICS: Lazy<BlockProducerMetrics> = Lazy::new(|| {
+    let metrics = BlockProducerMetrics::default();
+    let mut registry = gw_otel::metric::global();
+    metrics.register(registry.sub_registry_with_prefix("block_producer"));
+    metrics
+});
+
+pub static CUSTODIAN_METRICS: Lazy<CustodianMetrics> = Lazy::new(|| {
+    let mut metrics = CustodianMetrics::default();
+    debug_assert!(!metrics.enabled);
+
+    let maybe_enable = std::env::var(ENV_METRIC_MONITOR_CUSTODIAN_ENABLE);
+    if matches!(maybe_enable.as_deref(), Ok("true")) {
+        metrics.enabled = true;
+        let mut registry = gw_otel::metric::global();
+        metrics.register(registry.sub_registry_with_prefix("custodian"));
+    }
+
+    metrics
+});
+
+#[derive(Default)]
+pub struct BlockProducerMetrics {
+    pub resend: Counter,
+    pub witness_size: Counter,
+    pub tx_size: Counter,
+    pub sync_buffer_len: Gauge,
+    pub local_blocks: Gauge,
+    pub submitted_blocks: Gauge,
+}
+
+impl BlockProducerMetrics {
+    fn register(&self, registry: &mut Registry) {
+        registry.register(
+            "resend",
+            "Number of times resending submission transactions",
+            Box::new(self.resend.clone()),
+        );
+        registry.register_with_unit(
+            "witness_size",
+            "Block submission txs witness size",
+            Unit::Bytes,
+            Box::new(self.witness_size.clone()),
+        );
+        registry.register_with_unit(
+            "tx_size",
+            "Block submission txs size",
+            Unit::Bytes,
+            Box::new(self.tx_size.clone()),
+        );
+        registry.register(
+            "sync_buffer_len",
+            "Number of messages in the block sync receive buffer",
+            Box::new(self.sync_buffer_len.clone()),
+        );
+        registry.register(
+            "local_blocks",
+            "Number of local blocks",
+            Box::new(self.local_blocks.clone()),
+        );
+        registry.register(
+            "submitted_blocks",
+            "Number of submitted blocks",
+            Box::new(self.submitted_blocks.clone()),
+        );
```

Comment:
I still think the original approach (inline, fine grained metrics that lives close to the code that updates it) is superior.

Now metrics for read-only node and full-node are all mixed up, and the metrics doc is no longer accurate. Maybe we should remove it if the effort required to maintain it is not really justified.

Our crates are not exactly perfectly organized, e.g. the block sync client shouldn't really live in the block_producer crate, and the in-queue-request map and the request submitter maybe shouldn't live in the rpc server crate. Defining all metrics in crate level `metrics.rs` module results in weaker cohesion and undesirable coupling - if we want to move e.g. the block sync client to another crate, there will be a lot of churn dealing with its metrics.

Reply:
> I still think the original approach (inline, fine grained metrics that lives close to the code that updates it) is superior.

Original approach introduces at least four different code styles to integrate metrics.
- global structure
- use prometheus client type directly in business logic structure
- pass metric structure all the way down
- register new metric anywhere

To find out where this metric comes from is difficult. 

> Now metrics for read-only node and full-node are all mixed up, and the metrics doc is no longer accurate. Maybe we should remove it if the effort required to maintain it is not really justified.

Yes, should remove it. 
 
> Our crates are not exactly perfectly organized, e.g. the block sync client shouldn't really live in the block_producer crate, and the in-queue-request map and the request submitter maybe shouldn't live in the rpc server crate. Defining all metrics in crate level `metrics.rs` module results in weaker cohesion and undesirable coupling - if we want to move e.g. the block sync client to another crate, there will be a lot of churn dealing with its metrics.

That why my original plan is define all metrics in gw-metrics crate. Don't bother about metrics if we only want to move code into another crate. Everything still works.

Reply:
> > I still think the original approach (inline, fine grained metrics that lives close to the code that updates it) is superior.
> 
> Original approach introduces at least four different code styles to integrate metrics.

It's one style documented in `gw-metrics`:

```
//! Each crate/module/component can define their own metrics and register them
//! to the global `REGISTRY`. To avoid naming conflict, each of them SHOULD use
//! a unique prefix, e.g. the crate name or component name.
//!
//! If it makes sense to define a metrics struct, e.g. when there are a few
//! related metrics and they usually change together, it SHOULD live in a
//! separate metrics module. See the metrics module in gw-chain for an example.
```

>     * global structure

There are actually no global metrics structure after I moved `ChainMetrics` to `gw-chain`.

>     * use prometheus client type directly in business logic structure

It's recommended by the original style. By having a single source of truth, it's much less likely that metrics are not updated when business logic changes.

>     * pass metric structure all the way down

It's the context structure that's passed down, not a metrics structure. The metrics naturally belongs to the context.

>     * register new metric anywhere

It's recommended by the original style.

> To find out where this metric comes from is difficult. Update metrics for different ops is difficult.

Not really. Just search for the metrics name or description.

And why is updating metrics for different ops difficult?

> Now
> 
>     * crate metrics.rs and global structure
> 
> 
> > Now metrics for read-only node and full-node are all mixed up, and the metrics doc is no longer accurate. Maybe we should remove it if the effort required to maintain it is not really justified.
> 
> Yes, should remove it.
> 
> > Our crates are not exactly perfectly organized, e.g. the block sync client shouldn't really live in the block_producer crate, and the in-queue-request map and the request submitter maybe shouldn't live in the rpc server crate. Defining all metrics in crate level `metrics.rs` module results in weaker cohesion and undesirable coupling - if we want to move e.g. the block sync client to another crate, there will be a lot of churn dealing with its metrics.
> 
> That why my original plan is define all metrics in gw-metrics crate. Don't bother about metrics if we only want to move code into another crate. Everything still works.

What I had in mind is that we can have a metrics module under the e.g. `in_queue_request_map` module, if we have to define metrics in a metrics module. Not directly under the rpc-server crate root.

Reply:
I create a discussion, we can continue there.